File: src/main/java/uniks/accounting/segroup/Assignment.java
package uniks.accounting.segroup;

import java.beans.PropertyChangeSupport;

import java.beans.PropertyChangeListener;

public class Assignment  
{

   public static final String PROPERTY_points = "points";

   private double points;

   public double getPoints()
   {
      return points;
   }

   public Assignment setPoints(double value)
   {
      if (value != this.points)
      {
         double oldValue = this.points;
         this.points = value;
         firePropertyChange("points", oldValue, value);
      }
      return this;
   }


   private SEClass seClass = null;

   public SEClass getSeClass()
   {
      return this.seClass;
   }

   public Assignment setSeClass(SEClass value)
   {
      if (this.seClass != value)
      {
         SEClass oldValue = this.seClass;
         if (this.seClass != null)
         {
            this.seClass = null;
            oldValue.withoutAssignments(this);
         }
         this.seClass = value;
         if (value != null)
         {
            value.withAssignments(this);
         }
         firePropertyChange("seClass", oldValue, value);
      }
      return this;
   }



   public static final java.util.ArrayList<Solution> EMPTY_solutions = new java.util.ArrayList<Solution>()
   { @Override public boolean add(Solution value){ throw new UnsupportedOperationException("No direct add! Use xy.withSolutions(obj)"); }};


   private java.util.ArrayList<Solution> solutions = null;

   public java.util.ArrayList<Solution> getSolutions()
   {
      if (this.solutions == null)
      {
         return EMPTY_solutions;
      }

      return this.solutions;
   }

   public Assignment withSolutions(Object... value)
   {
      if(value==null) return this;
      for (Object item : value)
      {
         if (item == null) continue;
         if (item instanceof java.util.Collection)
         {
            for (Object i : (java.util.Collection) item)
            {
               this.withSolutions(i);
            }
         }
         else if (item instanceof Solution)
         {
            if (this.solutions == null)
            {
               this.solutions = new java.util.ArrayList<Solution>();
            }
            if ( ! this.solutions.contains(item))
            {
               this.solutions.add((Solution)item);
               ((Solution)item).setAssignment(this);
               firePropertyChange("solutions", null, item);
            }
         }
         else throw new IllegalArgumentException();
      }
      return this;
   }



   public Assignment withoutSolutions(Object... value)
   {
      if (this.solutions == null || value==null) return this;
      for (Object item : value)
      {
         if (item == null) continue;
         if (item instanceof java.util.Collection)
         {
            for (Object i : (java.util.Collection) item)
            {
               this.withoutSolutions(i);
            }
         }
         else if (item instanceof Solution)
         {
            if (this.solutions.contains(item))
            {
               this.solutions.remove((Solution)item);
               ((Solution)item).setAssignment(null);
               firePropertyChange("solutions", item, null);
            }
         }
      }
      return this;
   }


   protected PropertyChangeSupport listeners = null;

   public boolean firePropertyChange(String propertyName, Object oldValue, Object newValue)
   {
      if (listeners != null)
      {
         listeners.firePropertyChange(propertyName, oldValue, newValue);
         return true;
      }
      return false;
   }

   public boolean addPropertyChangeListener(PropertyChangeListener listener)
   {
      if (listeners == null)
      {
         listeners = new PropertyChangeSupport(this);
      }
      listeners.addPropertyChangeListener(listener);
      return true;
   }

   public boolean addPropertyChangeListener(String propertyName, PropertyChangeListener listener)
   {
      if (listeners == null)
      {
         listeners = new PropertyChangeSupport(this);
      }
      listeners.addPropertyChangeListener(propertyName, listener);
      return true;
   }

   public boolean removePropertyChangeListener(PropertyChangeListener listener)
   {
      if (listeners != null)
      {
         listeners.removePropertyChangeListener(listener);
      }
      return true;
   }

   public boolean removePropertyChangeListener(String propertyName,PropertyChangeListener listener)
   {
      if (listeners != null)
      {
         listeners.removePropertyChangeListener(propertyName, listener);
      }
      return true;
   }



   public void removeYou()
   {
      this.setCurrentGroup(null);
      this.setSeClass(null);
      this.setSolutionFolder(null);

      this.withoutSolutions(this.getSolutions().clone());


   }












   public static final String PROPERTY_task = "task";

   private String task;

   public String getTask()
   {
      return task;
   }

   public Assignment setTask(String value)
   {
      if (value == null ? this.task != null : ! value.equals(this.task))
      {
         String oldValue = this.task;
         this.task = value;
         firePropertyChange("task", oldValue, value);
      }
      return this;
   }


   @Override
   public String toString()
   {
      StringBuilder result = new StringBuilder();

      result.append(" ").append(this.getTask());


      return result.substring(1);
   }

   public static final String PROPERTY_seClass = "seClass";

   public static final String PROPERTY_solutions = "solutions";

   public static final String PROPERTY_solutionFolder = "solutionFolder";

   private SolutionFolder solutionFolder = null;

   public SolutionFolder getSolutionFolder()
   {
      return this.solutionFolder;
   }

   public Assignment setSolutionFolder(SolutionFolder value)
   {
      if (this.solutionFolder != value)
      {
         SolutionFolder oldValue = this.solutionFolder;
         if (this.solutionFolder != null)
         {
            this.solutionFolder = null;
            oldValue.setAssignment(null);
         }
         this.solutionFolder = value;
         if (value != null)
         {
            value.setAssignment(this);
         }
         firePropertyChange("solutionFolder", oldValue, value);
      }
      return this;
   }



   public static final String PROPERTY_currentGroup = "currentGroup";

   private SEGroup currentGroup = null;

   public SEGroup getCurrentGroup()
   {
      return this.currentGroup;
   }

   public Assignment setCurrentGroup(SEGroup value)
   {
      if (this.currentGroup != value)
      {
         SEGroup oldValue = this.currentGroup;
         if (this.currentGroup != null)
         {
            this.currentGroup = null;
            oldValue.setCurrentAssignment(null);
         }
         this.currentGroup = value;
         if (value != null)
         {
            value.setCurrentAssignment(this);
         }
         firePropertyChange("currentGroup", oldValue, value);
      }
      return this;
   }



}

File: typescript/SEGroup/libs/segroup-model/src/lib/Assignment.ts
import Solution from "./Solution";

import SEClass from "./SEClass";import SolutionFolder from "./SolutionFolder";import SEGroup from "./SEGroup";





  export default class Assignment  
{
    public task: string ;

    public points: number ;

    constructor() {
      this.task = '';
      this.points = 0;
      this._currentGroup = null;
      this._seClass = null;
      this._solutions = [];
      this._solutionFolder = null;

    }

    private _seClass: SEClass;

    get seClass(): SEClass {
      return this._seClass;
    }

    set seClass(value: SEClass) {
      if (this._seClass !== value) {
        const oldValue: SEClass = this._seClass;
        if (this._seClass) {
          this._seClass = null;
          oldValue.withoutAssignments(this);
        }
        this._seClass = value;
        if (value) {
          value.withAssignments(this);
        }
      }
    }



    private _solutions: Solution[];

    get solutions(): Solution[] {
      return this._solutions;
    }

    public withSolutions(...value: any[]): Assignment {
      if (!value) return this;

      for (const item of value) {
        if (!item) continue;
        if (item instanceof Array) {
          this.withSolutions(item);
        } else if (item instanceof Solution) {
          if (!this._solutions.includes(item)) {
            this._solutions.push(item);
            (item as Solution).assignment = this;
          }
        }
      }

      return this;
    }



    public withoutSolutions(...value: any[]): Assignment {
      if (this._solutions === [] || !value) return this;

      for (const item of value) {
        if (!item) continue;
        if (item instanceof Array) {
          this.withoutSolutions(...item);
        } else if (item instanceof Solution) {
          if (this._solutions.includes(item)) {
            this._solutions.splice(this._solutions.indexOf(item, 0), 1);
            (item as Solution).assignment = null;
          }
        }
      }
      return this;
    }


    public removeYou() {
      this.currentGroup = null;
      this.seClass = null;
      this.withoutSolutions(this._solutions);
      this.solutionFolder = null;

    }


    private _solutionFolder: SolutionFolder;

    get solutionFolder(): SolutionFolder {
      return this._solutionFolder;
    }

    set solutionFolder(value: SolutionFolder) {
      if (this._solutionFolder !== value) {
        const oldValue: SolutionFolder = this._solutionFolder;
        if (this._solutionFolder) {
          this._solutionFolder = null;
          oldValue.assignment = null;
        }
        this._solutionFolder = value;
        if (value) {
          value.assignment = this;
        }
      }
    }



    private _currentGroup: SEGroup;

    get currentGroup(): SEGroup {
      return this._currentGroup;
    }

    set currentGroup(value: SEGroup) {
      if (this._currentGroup !== value) {
        const oldValue: SEGroup = this._currentGroup;
        if (this._currentGroup) {
          this._currentGroup = null;
          oldValue.currentAssignment = null;
        }
        this._currentGroup = value;
        if (value) {
          value.currentAssignment = this;
        }
      }
    }



}